static/src/javascripts/projects/common/modules/commercial/membership-engagement-banner.js

```diff
@@ -109,31 +82,66 @@ const getUserVariantParams = (
  * otherwise a populated params object that looks like this:
  *
  *  {
- *    minArticles: 5, // how many articles should the user see before they get the engagement banner?
  *    messageText: "..."
  *    buttonCaption: "Become a Supporter"
  *  }
  *
  */
-const deriveBannerParams = (location: string): ?EngagementBannerParams => {
-    const defaultParams = engagementBannerParams(location);
+
+const buildCampaignCode = (
+    userTest: ?AcquisitionsABTest,
+    userVariant: ?Variant
+): ?{ campaignCode: string } => {
+    if (userTest && userVariant) {
+        const params = userVariant.engagementBannerParams;
+        if (params && params.campaignCode) {
+            return params.campaignCode;
+        }
+        return { campaignCode: `${userTest.campaignId}_${userVariant.id}` };
+    }
+};
+
+const deriveBannerParams = (): Promise<?EngagementBannerParams> => {
+    const userTest: ?AcquisitionsABTest = getUserTest();
+    const userVariant: ?Variant = getUserVariant(userTest);
+    const defaultParams: EngagementBannerParams = defaultEngagementBannerParams();
+
+    // if the user isn't in a test, use the control in google docs
+    if (!userVariant) {
+        return getControlEngagementBannerParams().then(controlParams => ({
+            ...defaultParams,
+            ...controlParams,
+        }));
+    }
+
+    if (userVariant && userVariant.engagementBannerParams) {
+        const campaignCode: ?{ campaignCode: string } = buildCampaignCode(
+            userTest,
+            userVariant
+        );
+
+        return getUserVariantParams(userVariant).then(variantParams => ({
+            ...defaultParams,
+            ...variantParams,
+            ...campaignCode,
```

Comment:
could we maybe pull this out into a function, i.e `getCampaignCode(userVariant)`? There's quite a lot of logic happening in this function and I think this would make it more readable